store/sqlstore/migration.go

```diff
@@ -0,0 +1,194 @@
+	Execute(context.Context, SqlStore, *sql.Conn) (AsyncMigrationStatus, error)
+}
+
+// MigrationRunner runs queued async migrations
+type MigrationRunner struct {
+	supplier   SqlStore
+	migrations []AsyncMigration
+	done       chan struct{}
+	options    MigrationOptions
+	cancelFn   context.CancelFunc
+}
+
+type MigrationOptions struct {
+	LockTimeout int // Lock timeout in seconds
+	BackoffTime int // Time to wait between retries
+	NumRetries  int // Number of retries in case of failed migration
+}
+
+func NewMigrationRunner(s SqlStore, o MigrationOptions) *MigrationRunner {
+	if o.LockTimeout == 0 {
+		o.LockTimeout = MigrationDefaultLockTimeout
+	}
+	if o.BackoffTime == 0 {
+		o.BackoffTime = MigrationDefaultBackoffTime
+	}
+	if o.NumRetries == 0 {
+		o.NumRetries = MigrationDefaultNumRetries
+	}
+	return &MigrationRunner{
+		supplier: s,
+		done:     make(chan struct{}),
+		options:  o,
+	}
+}
+
+// Add checks if the migration should be executed and adds it to queue
+func (r *MigrationRunner) Add(m AsyncMigration) error {
+	// check status in Systems table
+	currentStatus, appErr := r.supplier.System().GetByName("migration_" + m.Name())
+	if appErr == nil {
+		if currentStatus.Value == string(AsyncMigrationStatusComplete) || currentStatus.Value == string(AsyncMigrationStatusSkip) {
+			return nil
+		}
+	}
+	// get status from migration
+	status, err := m.GetStatus(r.supplier)
+	if err != nil {
+		return err
+	}
+	if status == AsyncMigrationStatusComplete || status == AsyncMigrationStatusSkip {
+		return nil
+	}
+	r.migrations = append(r.migrations, m)
+	return nil
+}
+
+// Run all queued migrations sequentially
+func (r *MigrationRunner) Run() error {
+	go func() {
+		ctx, cancelFn := context.WithCancel(context.Background())
+		defer cancelFn()
+		defer close(r.done)
+		r.cancelFn = cancelFn
+
+		for _, m := range r.migrations {
+			// function that will try to execute migration
+			migrate := func() error {
+				var err error
+				conn, err := createConnectionWithLockTimeout(ctx, r.supplier, r.options.LockTimeout)
+				if err != nil {
+					mlog.Error("Failed to setup connection", mlog.Err(err))
+					return err
+				}
+				defer releaseConnection(ctx, r.supplier, conn)
+
+				// run migration
+				status, err := m.Execute(ctx, r.supplier, conn)
+				if err != nil {
+					mlog.Error("Failed to execute migration", mlog.Err(err))
+					return err
+				}
+				if status == AsyncMigrationStatusComplete || status == AsyncMigrationStatusSkip {
+					// save migration status
+					r.supplier.System().SaveOrUpdate(&model.System{Name: "migration_" + m.Name(), Value: string(status)})
+				} else if status == AsyncMigrationStatusFailed {
+					mlog.Error("Failed migration " + m.Name())
+					return errors.New("Failed migration")
+				} else {
+					// should we return error here to retry?
+					return errors.New("Invalid result from migration")
+				}
+				return nil
+			}
+			// retry migration if it fails
```

Comment:
I'm a bit wary about this. Are all migrations guaranteed to be idempotent ? I am wondering whether we will get into a more messed up state trying to execute a failed migration again.

I don't know much about mysql, but postgres says:

> If a problem arises while scanning the table, such as a deadlock or a uniqueness violation in a unique index, the CREATE INDEX command will fail but leave behind an “invalid” index. This index will be ignored for querying purposes because it might be incomplete; however it will still consume update overhead. The psql \d command will report such an index as INVALID:
The recommended recovery method in such cases is to drop the index and try again to perform CREATE INDEX CONCURRENTLY. (Another possibility is to rebuild the index with REINDEX INDEX CONCURRENTLY).

Will it be safer to execute "REINDEX INDEX CONCURRENTLY" for all tries after the first one ? And leave "CREATE INDEX CONCURRENTLY" for the first one.

This may need changes to the interface to add another method for all tries after the first one.

Reply:
That's a good point. `GetStatus` should also check if the index is valid. In `Execute` that would also have to be checked and invalid index dropped first (reindex concurrently is only in postgres 12).

Reply:
> (reindex concurrently is only in postgres 12).

Ah okay. Let's do the checking and dropping then.